### ISSUE TITLE
Question: How to re-use same transaction across multiple threads?

### ISSUE THREAD

**Issue by sgzsh269** [ebean-orm/ebean#1931]
I am trying to figure out a way to re-use same transaction across multiple threads to allow related  objects to be saved together for the purpose of atomicity. 
I have been exploring the Transaction API and found `createTransaction()` that is not thread-local and can be passed around
Is the use of `createTransaction()` only way to re-use transaction across threads or are there alternatives? 
Any insight would help. Thanks.

**Comment by rbygrave** [comment]
*> Is the use of createTransaction() only way to re-use transaction across
threads or are there alternatives?*

It's the only way currently.  So using createTransaction() which will
"work" and is "ok" but style wise this code is different from all the other
code in terms of style (explicitly passing the transaction around for
queries and persisting which we otherwise don't do - so this code will look
different style wise).

Note that For the "Future" Queries like findFutureList() etc Ebean
internally moves the transactions to the other thread - I'm sure that isn't
useful for the case we are talking about here.



I *suspect* what we'd really like for this case is:

// detach the transaction from ThreadLocal - called before we move the rest
of the work to another thread
transaction.detach()

// attach the transaction to the ThreadLocal - called when work with
transaction is moved to the new thread
transaction.attach();

-----------------------

So then we could do something like:

Transaction txn = DB.beginTransaction();
try {
  // do some initial work in the foreground, maybe lock a row
  new QJob().name.eq("myjobId").forUpdate().findOne();

  // successfully done some initial work in foreground
  // and we want the rest to execute in background

  // 1. first detach transaction from TheadLocal
  txn.detach();

  // 2. now kick off rest of task in background passing the transaction
  executor.runInBackground(() -> {
    runWithTxnInBackground(txn);
  });

} catch (Exception e) {
  // handle failure of initial work
  txn.rollback();
}

void runWithTxnInBackground(Transaction txn) {
  try {
    // attach transaction to the ThreadLocal
    txn.attach();

    // do stuff here in background...

    txn.commit();
  } catch (Exception e) {
    log.error("background task failed ... ", e);
  } finally {
    txn.end();
  }
}


Does any of that make sense ?

Cheers, Rob.

**Comment by jnehlmeier** [comment]
I am curious: If you talk about detach/attach to move transaction from one thread to another, does that mean an Ebean transaction can generally only be owned by single thread at any time? So using something like the fork/join framework / Scheduler within a transaction isn't possible?

**Comment by rbygrave** [comment]
*> does that mean an Ebean transaction can generally only be owned by
single thread at any time*

Depends on wording here so no there is not a *strict* requirement for a
transaction to be owned by a single thread.  In JDBC this maps to a
java.sql.Connection being allowed to be used by multiple threads.

However, *practically* speaking most of the time we are
using @Transactional or DB.beginTransaction() and these attach the
transaction to a ThreadLocal.  A ThreadLocal is by nature local to the
single current thread.  When we do this we don't need to "pass the
transaction around" to anything and everything that uses it - pass it
between threads or amongst threads.

When we use DB.createTransaction() ... then naturally we are in the
situation where we need to pass the transaction around explicitly and use
it explicitly.



*> So using something like the fork/join framework / Scheduler within a
transaction isn't possible?*

Can you expand this statement and be more explicit about what you mean?  Do
you have cases of sharing a transaction across multiple concurrently
executing threads? I'd be keen to hear because I can't remember when I last
did that.

In my opinion the common case is to pass a transaction from one thread to
another (often foreground to background) and there isn't any magic in that
we need to pass it around explicitly.


-----------------------
Background note:

I presume people know about JVM Project Loom bringing "Virtual Threads" to
the JVM but obviously this is a HUGE DEAL from a JDBC perspective as it
effectively means all Type 4 JDBC drivers become non-blocking (in fact all
Java IO becomes non-blocking). So this likely compares favourably to the
world of "async jdbc drivers" (where transactions are passed as part of the
callback scope).

The "pressure" on @Transactional / ThreadLocal passing comes from the
desire for lower resource consumption (Thread memory and JDBC Blocking
API).  Hence the Reactive/Async style and associated drivers designed to
replace JDBC. With Loom coming JDBC will become non-blocking and consume
less memory via "Virtual Threads" so we should get all the benefits of the
Reactive/Async approach but without any code or design changes (Loom ->
code like sync, perform like async).

**Comment by rbygrave** [comment]
*More Background - Loom's impact on Ebean:*

So without Loom coming to the JVM Ebean would be somewhat conflicted in
terms on what or how much it should support async jdbc drivers (like r2dbc
etc).  With Loom coming it is now very likely that Ebean doesn't support
any of the [proprietary] async drivers (which would be non trivial as they
use database specific bind params etc).  Loom brings the same benefits to
users of Ebean without any change required (as long as we are using Type 4
JDBC drivers).  In a similar way this is also why the efforts for a
standardised async driver API has been abandoned [by Oracle and company] -
but sure the r2dbc folks probably disagree and Loom isn't in an officially
released JDK yet.

For me, what also sometimes gets lost, forgotten about or under stated in
the Loom / async driver resource discussion is that the *database shadow
process* associated with the connection is a relatively more important
resource than our client/app server thread/memory.

**Comment by jnehlmeier** [comment]
> When we use DB.createTransaction() ... then naturally we are in the situation where we need to pass the transaction around explicitly and use it explicitly.

Maybe you could update the Ebean APIs a bit so you can use it more straightforward, e.g.

```java
new QPerson(tx).id.eq(1).findOne();
// could delegate to tx.save(this) similar to Model.save() delegating 
// to db().save(this). db() would figure out the current TX based on 
// ThreadLocal and delegate to currentTx().save(..).
person.save(tx);
```

Maybe it is worth creating a common interface related to query execution that is implemented by Database and Transaction. The difference would be that Database would use a ThreadLocal transaction internally and delegate to it while Transaction would do the actual work.
That would allow you to remove all the methods on Database that require a transaction parameter, since you could execute them directly on the Transaction instance. And APIs that currently require a Database instance could also take a Transaction instance (see above code example).

**Comment by rbygrave** [comment]
## Query beans
```java
new QPerson(tx).id.eq(1).findOne();
```
Yes, that looks nice and intuitive.  Note that right now for ORM queries there are `usingTransaction(Transaction txn)` and `usingConnection(Connection connection)` methods so right now we can instead do:
```java
new QPerson().usingTransaction(tx).id.eq(1).findOne();
```
... but yes it might be more obvious if this was a constructor option for query beans.

In terms of DtoQuery I note that we are missing this option/ability on DtoQuery so that looks like a gap/issue there.  We can fix that using the same approach used in ORM queries and adding `usingTransaction(Transaction txn)` and `usingConnection(Connection connection)` methods .


## Model 
```java
person.save(tx);
```
Yes, that is an easy almost trivial addition to the Model API for save, delete, insert, update. 


> creating a common interface ... that is implemented by Database and Transaction.

It's an idea and we should not dismiss it yet but I think we can achieve our goals easily enough without doing that [_using query beans and model with explicit transactions_]. Note that with Ebean we make use of "implicit read only transactions" and support using "external transactions" so the common interface approach has some potential challenges.  We can go with the easy option first.

**Comment by rbygrave** [comment]
- DtoQuery.usingTransaction - https://github.com/ebean-orm/ebean/issues/1933

**Comment by rbygrave** [comment]
@sgzsh269 Did the above make sense?  That satisfies what you are looking to do right?

**Comment by sgzsh269** [comment]
@rbygrave thanks for taking the time to provide your insight, it is really helpful. 
My use-case primarily is to re-use a transaction from one thread to another and not really have concurrent threads use the same transaction. I am using Scala Future to orchestrate the threads

`createTransaction()` may not have been created for such a purpose, but if you say it is 'ok' I could try it out and see if it works for me.

**Comment by rbygrave** [comment]
> re-use a transaction from one thread to another

Yes cool. As I see it this is be far the most common case.


> `createTransaction()` may not have been created for such a purpose

`createTransaction()` was created for this purpose so absolutely use it. The javadoc isn't as clear on that point as it should be.

As I see it the issues/discussion above is all largely about the **elegance** of the Ebean API for using an explicit transaction and the improvements required to make it nicer. 
- model.save(transaction) 
- query bean constructor taking transaction - `new QPerson(tx)`
- DtoQuery.usingTransaction(transaction)

That is, this functionality was on the Database API [and ExtendedServer database.extended()] ... per say but was not as elegant / nice as we want.

Note that all those enhancements will be in the next release - 12.1.12.

**Comment by rbygrave** [comment]
Ok, I think we can close this for now.  Re-open if there are further questions on this.
